Tests/NIOExtrasTests/LengthFieldBasedFrameDecoderTest.swift

```diff
@@ -0,0 +1,298 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import XCTest
+import NIO
+import NIOExtras
+
+private let standardDataString = "abcde"
+
+class LengthFieldBasedFrameDecoderTest: XCTestCase {
+    
+    private var channel: EmbeddedChannel!
+    private var decoderUnderTest: LengthFieldBasedFrameDecoder!
+    
+    override func setUp() {
+        channel = EmbeddedChannel()
+    }
+
+    func testDecodeWithUInt8HeaderWithData() throws {
+        
+        decoderUnderTest = LengthFieldBasedFrameDecoder(lengthFieldLength: 1)
+        try? channel.pipeline.add(handler: decoderUnderTest).wait()
+        
+        let dataBytes: [UInt8] = [10, 20, 30, 40]
+        let dataBytesLengthHeader = UInt8(dataBytes.count)
+        
+        var buffer = channel.allocator.buffer(capacity: 5)
+        buffer.write(bytes: [dataBytesLengthHeader])
+        buffer.write(bytes: dataBytes)
+        
+        XCTAssertTrue(try channel.writeInbound(buffer))
+        
+        var outputBuffer: ByteBuffer? = channel.readInbound()
+        let outputData = outputBuffer?.readBytes(length: dataBytes.count)
+        
+        XCTAssertEqual(dataBytes, outputData)
+        XCTAssertFalse(try channel.finish())
+    }
+    
+    func testDecodeWithUInt8HeaderWithString() throws {
+        
+        decoderUnderTest = LengthFieldBasedFrameDecoder(lengthFieldLength: 1)
+        try? channel.pipeline.add(handler: decoderUnderTest).wait()
+        
+        var dataLength: UInt8 = 5
+        
+        let headerData = Data(bytes: &dataLength, count: MemoryLayout.size(ofValue: dataLength))
+        
+        var buffer = channel.allocator.buffer(capacity: 6)
+        buffer.write(bytes: headerData)
+        buffer.write(string: standardDataString)
+        
+        XCTAssertTrue(try channel.writeInbound(buffer))
+        
+        var outputBuffer: ByteBuffer? = channel.readInbound()
+        let outputData = outputBuffer?.readString(length: standardDataString.count)
+        
+        XCTAssertEqual(standardDataString, outputData)
+        XCTAssertFalse(try channel.finish())
+    }
+    
+    func testDecodeWithUInt16HeaderWithString() throws {
+        
+        decoderUnderTest = LengthFieldBasedFrameDecoder(lengthFieldLength: 2)
+        try? channel.pipeline.add(handler: decoderUnderTest).wait()
+        
+        var dataLength: UInt16 = 5
+        
+        let headerData = Data(bytes: &dataLength, count: MemoryLayout.size(ofValue: dataLength))
```

Comment:
This test is very subtly wrong if run on platforms that are not little-endian. This is because the default behaviour of the `LengthFieldBasedFrameDecoder` is to decode integers in a little-endian representation, and if the platform in question is big-endian then the integer will be incorrectly decoded.

Rather than use `Data` as a temporary intermediary, why not use `ByteBuffer.write(integer:)` to write the `UInt16`? That will allow you to ensure the the write uses the same endianness as the `LengthFieldBasedFrameDecoder`.

Incidentally, there are no tests that exercise the endianness parameter: perhaps there should be.

Reply:
Sure, all good points. Will change.